### PR TITLE
gh-112730: Make the test suite resilient to color-activation environment variables

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-112730

### CODE REVIEW

File: .github/workflows/reusable-ubuntu.yml [hunk 1]
```diff
@@ -16,6 +16,7 @@ jobs:
     env:
       OPENSSL_VER: 3.0.13
       PYTHONSTRICTEXTENSIONBUILD: 1
+      FORCE_COLOR: 1
```
Comment (hugovk) [inline .github/workflows/reusable-ubuntu.yml]: Nit: alphabetise, makes it easier for when it grows:
```suggestion
      FORCE_COLOR: 1
      OPENSSL_VER: 3.0.13
      PYTHONSTRICTEXTENSIONBUILD: 1
```

File: Lib/test/support/__init__.py [hunk 2]
```diff
@@ -2555,3 +2556,22 @@ def copy_python_src_ignore(path, names):
...
+            for key in variables:
+                variables[key] = os.environ.pop(key, None)
+            traceback._can_colorize = lambda: False
+            return func(*args, **kwargs)
+        finally:
+            traceback._can_colorize = original_fn
+            for key, value in variables.items():
+                if value is not None:
+                    os.environ[key] = value
+    return wrapper
```
Comment (hugovk) [inline Lib/test/support/__init__.py]: Lint fix:
```suggestion
    return wrapper

```

File: Lib/test/test_traceback.py [hunk 3]
```diff
@@ -4360,7 +4369,8 @@ def test_colorized_detection_checks_for_environment_variables(self):
         else:
             virtual_patching = contextlib.nullcontext()
         with virtual_patching:
-            with unittest.mock.patch("os.isatty") as isatty_mock:
+            with (unittest.mock.patch("os.isatty") as isatty_mock,
```
Comment (hugovk) [inline Lib/test/test_traceback.py]: Lint fix:
```suggestion
            with (unittest.mock.patch("os.isatty") as isatty_mock,
```

File: Lib/test/support/__init__.py [hunk 4]
```diff
@@ -58,7 +58,8 @@
     "LOOPBACK_TIMEOUT", "INTERNET_TIMEOUT", "SHORT_TIMEOUT", "LONG_TIMEOUT",
     "Py_DEBUG", "EXCEEDS_RECURSION_LIMIT", "Py_C_RECURSION_LIMIT",
     "skip_on_s390x",
-    "without_optimizer",
+    "without_optimizer", "set_not_colorized_globally", "reset_colorized_globally",
```
Comment (hugovk) [inline Lib/test/support/__init__.py]: Doesn't exist and not used anywhere? It fails `test_all` 

```suggestion
    "without_optimizer", "reset_colorized_globally",
```

File: Lib/traceback.py [hunk 5]
```diff
@@ -141,11 +141,12 @@ def _can_colorize():
                 return False
         except (ImportError, AttributeError):
             return False
-
-    if os.environ.get("PYTHON_COLORS") == "0":
-        return False
-    if os.environ.get("PYTHON_COLORS") == "1":
-        return True
+
```
Comment (hugovk) [inline Lib/traceback.py]: Lint fix:
```suggestion

```

**Comment by gvanrossum** [comment]
I am kind of sad that the solution is to add a special hack to suppress color to every test that would break. My ideal solution would be to suppress the color variables when `-E` is used, or for the traceback code to be able to detect whether `-E` is used.

**Comment by pablogsal** [comment]
> I am kind of sad that the solution is to add a special hack to suppress color to every test that would break. My ideal solution would be to suppress the color variables when `-E` is used, or for the traceback code to be able to detect whether `-E` is used.

That only takes half of the tests because that fixes the cases where the traceback is taken from a subprocess. The other half which is when the traceback is taken from the process that runs test itself you need the decorator because the test suite cannot be executed itself always with -E

**Comment by gvanrossum** [comment]
Got it. I wonder if `_can_colorize()` is also a bit naive? Its default is to checking whether stderr is a tty, even when its caller is going to write to some other stream.

**Comment by hugovk** [comment]
That came up for termcolor and we replaced:

```python
    return (
        hasattr(sys.stdout, "isatty")
        and sys.stdout.isatty()
        and os.environ.get("TERM") != "dumb"
    )
```

With:

```python
    # Then check system:
    if os.environ.get("TERM") == "dumb":
        return False
    if not hasattr(sys.stdout, "fileno"):
        return False

    try:
        return os.isatty(sys.stdout.fileno())
    except io.UnsupportedOperation:
        return sys.stdout.isatty()
```

https://github.com/termcolor/termcolor/pull/56/files

File: Lib/test/support/__init__.py [hunk 6]
```diff
@@ -58,7 +58,8 @@
     "LOOPBACK_TIMEOUT", "INTERNET_TIMEOUT", "SHORT_TIMEOUT", "LONG_TIMEOUT",
     "Py_DEBUG", "EXCEEDS_RECURSION_LIMIT", "Py_C_RECURSION_LIMIT",
     "skip_on_s390x",
-    "without_optimizer",
+    "without_optimizer", "reset_colorized_globally",
```
Comment (hugovk) [inline Lib/test/support/__init__.py]: Also not used anywhere and fails `test___all__`:
```suggestion
    "without_optimizer",
```

File: Lib/test/test_traceback.py [hunk 7]
```diff
@@ -387,7 +396,7 @@ def test_encoded_file(self):
         process = subprocess.Popen([sys.executable, "-c",
                                     "import sys; print(sys.stdout.encoding)"],
                                    stdout=subprocess.PIPE,
-                                   stderr=subprocess.STDOUT)
+                                   stderr=subprocess.STDOUT, env={})
```
Comment (gpshead) [inline Lib/test/test_traceback.py]: On some platforms I believe things can fail to execute at all with an empty environment missing some required information...?   should these tests use `test_support.script_helper.run_python_until_end()`?

File: Lib/test/test_cmd_line_script.py
Comment (gpshead) [inline Lib/test/test_cmd_line_script.py:1]: There's enough decorators in this test file I suggest just disabling it via setUpModule / tearDownModule instead.

**Review by gpshead — COMMENTED** [review]
So long as the number of decorator applications isn't huge I think this works.  We should pay attention to users during the alpha and beta cycle.  I'm assuming they aren't likely to find issues with this in CI and they probably do not enable colors there.  It feels more likely to happen to an end user developer working within their local dev environment (which I believe is how we found it amongst ourselves?)

**Comment by gpshead** [comment]
FWIW I do not mind if we just wind up disabling it entirely on `-E`, especially for the initial release of the feature. I agree that regardless of a PYTHON_  environment variable name prefix, it is a natural behavior change for people to look to -E and expect it to do.

**Comment by pablogsal** [comment]
Ok, I will move to disable on `-E` then and remove as many decorators as possible.

File: Lib/test/test_traceback.py (hunk 7 above)
Comment (pablogsal) [reply Lib/test/test_traceback.py]: Now that I am deactivating on `-E` this change is not needed anymore.

**Comment by hugovk** [comment]
This test is failing:

```pytb
FAIL: test_colorized_detection_checks_for_environment_variables (test.test_traceback.TestColorizedTraceback.test_colorized_detection_checks_for_environment_variables)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\test_traceback.py", line 4376, in test_colorized_detection_checks_for_environment_variables
    self.assertEqual(traceback._can_colorize(), False)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: True != False
```

Does it also need `@force_not_colorized`?

**Comment by kulikjak** [comment]
Hi. With this change, I am seeking several failures like the following one in `test_traceback` and `test_tracemalloc`: 

```
======================================================================
FAIL: test_env_var_invalid (test.test_tracemalloc.TestCommandLine.test_env_var_invalid) (nframe=-1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/..../pythonmain/cpython-main/Lib/test/test_tracemalloc.py", line 959, in test_env_var_invalid
    self.check_env_var_invalid(nframe)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
  File "/..../pythonmain/cpython-main/Lib/test/test_tracemalloc.py", line 953, in check_env_var_invalid
    self.fail(f"unexpected output: {stderr!a}")
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: unexpected output: b'ld.so.1: python: fatal: libpython3.13.so.1.0: open failed: No such file or directory\nld.so.1: python: fatal: relocation error: file /..../pythonmain/build/sparcv9/python: symbol Py_BytesMain: referenced symbol not found\n'
```
It seems that because the environment is now being cleared in those tests, `LD_LIBRARY_PATH` is no longer set in the subprocess, and `libpython3.13.so.1.0` thus not found.

I am seeing this on Oracle Solaris, though atm I don't think this is specific to us? Also, I am executing the test suite in a pretty standard way:

`cd /..../pythonmain/build/sparcv9; EXTRATESTOPTS="-v test_traceback test_tracemalloc" /usr/gnu/bin/make test`

**Comment by pablogsal** [comment]
I will look into this today!

**Comment by ericsnowcurrently** [comment]
FWIW, I've seen failures in test_traceback/test_tracemalloc on a number of stable buildbots since this change.

**Pull request merged**